### PR TITLE
feat(vmop): first implementation

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
 Implementation resource VirtualMachineOperation. 
 The VirtualMachineOperation custom resource allows for declarative management of a working virtual machine. It enables actions such as stopping, starting, and restarting the virtual machine.

## Why do we need it, and what problem does it solve?
The VirtualMachineOperation resource is needed to provide a high-level abstraction for managing virtual machines. It solves the problem of efficiently and easily controlling the state of virtual machines through a declarative approach, rather than relying on manual, time-consuming operations.

## What is the expected result?
VirtualMachineOperation is an immutable resource. After deployment, it can have 4 phases of status: Completed, Failed, InProgress, Pending.

### CODE REVIEW

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-controller/pkg/common/kvvm/util.go [hunk 1]
```diff
@@ -22,3 +29,133 @@ func GetRunStrategy(kvvm *virtv1.VirtualMachine) virtv1.VirtualMachineRunStrateg
...
+	return nil
+}
+
+// RestartKVVM restarts kvvm.
+func RestartKVVM(ctx context.Context, cli client.Client, kvvm *virtv1.VirtualMachine, kvvmi *virtv1.VirtualMachineInstance, force bool) error {
+	if kvvm == nil {
+		return fmt.Errorf("kvvm must not be empty")
+	}
+	if kvvmi == nil {
+		return fmt.Errorf("kvvm must not be empty")
```
Comment (Isteb4k) [inline images/virtualization-controller/pkg/common/kvvm/util.go:143]: **kvvmi** must not be empty

**Review by Isteb4k — APPROVED** [review]

File: images/virtualization-controller/pkg/controller/vmop/vmop_webhook.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		log: log.WithName(controllerName).WithValues("webhook", "validation"),
+	}
+}
+
+type Validator struct {
+	log logr.Logger
+}
+
+func (v *Validator) ValidateCreate(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	err := fmt.Errorf("misconfigured webhook rools: delete operation not implemented")
```
Comment (diafour) [inline images/virtualization-controller/pkg/controller/vmop/vmop_webhook.go:26]:
```suggestion
	err := fmt.Errorf("misconfigured webhook rules: delete operation not implemented")
```

File: images/virtualization-controller/pkg/controller/vmop/vmop_webhook.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+	if reflect.DeepEqual(oldVmop.Spec, newVmop.Spec) {
+		return nil, nil
+	}
+	err := fmt.Errorf("vmop %q is invalid. vmop.spec is immutable", oldVmop.GetName())
+	return nil, err
+}
+
+func (v *Validator) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	err := fmt.Errorf("misconfigured webhook rools: delete operation not implemented")
```
Comment (diafour) [inline images/virtualization-controller/pkg/controller/vmop/vmop_webhook.go:51]:
```suggestion
	err := fmt.Errorf("misconfigured webhook rules: delete operation not implemented")
```

File: images/virtualization-controller/pkg/controller/vmop/vmop_reconciler_state.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	state.inProgress = true
+}
+
+func (state *ReconcilerState) GetInProgress() bool {
+	return state.inProgress
+}
+
+func (state *ReconcilerState) GetKVVM(ctx context.Context) (*virtv1.VirtualMachine, error) {
+	if state.VmIsEmpty() {
+		return nil, fmt.Errorf("VM %s not found", state.VMOP.Current().Spec.VirtualMachineName)
```
Comment (diafour) [inline images/virtualization-controller/pkg/controller/vmop/vmop_reconciler_state.go:174]:
```suggestion
		return nil, fmt.Errorf("KVVM %s not found", state.VMOP.Current().Spec.VirtualMachineName)
```

File: images/virtualization-controller/pkg/controller/vmop/vmop_reconciler_state.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	kvvm := &virtv1.VirtualMachine{}
+	key := types.NamespacedName{Name: state.VM.GetName(), Namespace: state.VM.GetNamespace()}
+	err := state.Client.Get(ctx, key, kvvm)
+	return kvvm, err
+}
+
+func (state *ReconcilerState) GetKVVMI(ctx context.Context) (*virtv1.VirtualMachineInstance, error) {
+	if state.VmIsEmpty() {
+		return nil, fmt.Errorf("VM %s not found", state.VMOP.Current().Spec.VirtualMachineName)
```
Comment (diafour) [inline images/virtualization-controller/pkg/controller/vmop/vmop_reconciler_state.go:184]:
```suggestion
		return nil, fmt.Errorf("KVVMI %s not found", state.VMOP.Current().Spec.VirtualMachineName)
```

File: images/virtualization-controller/pkg/controller/vmop/vmop_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+func (r *Reconciler) doOperation(ctx context.Context, operationSpec virtv2.VirtualMachineOperationSpec, state *ReconcilerState) error {
+	switch operationSpec.Type {
+	case virtv2.VMOPOperationTypeStart:
+		return r.doOperationStart(ctx, state)
+	case virtv2.VMOPOperationTypeStop:
+		return r.doOperationStop(ctx, operationSpec.Force, state)
+	case virtv2.VMOPOperationTypeRestart:
+		return r.doOperationRestart(ctx, operationSpec.Force, state)
+	default:
+		return fmt.Errorf("unexpected opearation %q. %w", operationSpec.Type, common.ErrUnknownValue)
```
Comment (diafour) [inline images/virtualization-controller/pkg/controller/vmop/vmop_reconciler.go:172]:
```suggestion
		return fmt.Errorf("unexpected operation %q. %w", operationSpec.Type, common.ErrUnknownValue)
```

File: images/virtualization-controller/pkg/common/kvvm/util.go [hunk 7]
```diff
@@ -22,3 +29,133 @@ func GetRunStrategy(kvvm *virtv1.VirtualMachine) virtv1.VirtualMachineRunStrateg
...
+// DeletePodByKVVMI deletes pod by kvvmi.
+func DeletePodByKVVMI(ctx context.Context, cli client.Client, kvvmi *virtv1.VirtualMachineInstance, opts client.DeleteOption) error {
+	pod, err := FindPodByKVVMI(ctx, cli, kvvmi)
+	if err != nil {
+		return err
+	}
+	if pod == nil {
+		return nil
+	}
+	return cli.Delete(ctx, pod, opts)
```
Comment (diafour) [inline images/virtualization-controller/pkg/common/kvvm/util.go:77]: Should handle IsNotFound error

**Pull request merged**